homeassistant/components/lock/august.py

```diff
@@ -0,0 +1,97 @@
+"""
+Support for August lock.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/lock.august/
+"""
+from datetime import timedelta
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.august import DATA_AUGUST, DOMAIN
+from homeassistant.components.lock import LockDevice, ATTR_CHANGED_BY
+from homeassistant.const import ATTR_BATTERY_LEVEL, CONF_SCAN_INTERVAL
+
+DEPENDENCIES = ['august']
+
+SCAN_INTERVAL = timedelta(seconds=5)
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_SCAN_INTERVAL, default=SCAN_INTERVAL): cv.time_period
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+ATTR_SERIAL_NUMBER = "serial_number"
+ATTR_FIRMWARE_VERSION = "firmware_version"
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up August locks."""
+    data = hass.data[DATA_AUGUST]
+    devices = []
+
+    for lock in data.locks:
+        devices.append(AugustLock(data, lock))
+
+    add_devices(devices, True)
+
+
+class AugustLock(LockDevice):
+    """Representation of an Abode lock."""
+
+    def __init__(self, data, lock):
+        """Initialize the lock."""
+        self._data = data
+        self._lock = lock
+        self._lock_status = None
+        self._lock_detail = None
+        self._changed_by = None
+
+    def lock(self, **kwargs):
+        """Lock the device."""
+        self._data.lock(self._lock.device_id)
+
+    def unlock(self, **kwargs):
+        """Unlock the device."""
+        self._data.unlock(self._lock.device_id)
+
+    def update(self):
+        """Get the latest state of the sensor."""
+        self._lock_status = self._data.get_lock_status(self._lock.device_id)
+        self._lock_detail = self._data.get_lock_detail(self._lock.device_id)
+
+        from august.activity import ActivityType
+        activity = self._data.get_latest_device_activity(
+            self._lock.device_id,
+            ActivityType.LOCK_OPERATION)
+
+        if activity is not None:
+            self._changed_by = activity.operated_by
+
+    @property
+    def name(self):
+        """Return the name of this device."""
+        return self._lock.device_name
+
+    @property
+    def is_locked(self):
+        """Return true if device is on."""
+        from august.lock import LockStatus
+        return self._lock_status is LockStatus.LOCKED
+
+    @property
+    def changed_by(self):
+        """Last change triggered by."""
+        return self._changed_by
+
+    @property
+    def device_state_attributes(self):
+        """Return the device specific state attributes."""
+        return {
+            ATTR_BATTERY_LEVEL: "{}%".format(self._lock_detail.battery_level),
+            ATTR_SERIAL_NUMBER: self._lock_detail.serial_number,
+            ATTR_FIRMWARE_VERSION: self._lock_detail.firmware_version,
+            ATTR_CHANGED_BY: self._changed_by
```

Comment:
Hmm, the base lock entity class seems to do a bad thing in the `state_attributes` property. It's only allowing state attributes to be reported if the `code_format` property returns something. That's not good. I think we should fix this. It should return all attributes that are not `None`.
```py
PROP_TO_ATTR = {
    'changed_by': ATTR_CHANGED_BY,
    'code_format': ATTR_CODE_FORMAT,
}
...
@property
def state_attributes(self):
    """Return the state attributes."""
    data = {}
    for prop, attr in PROP_TO_ATTR.items():
        value = getattr(self, prop)
        if value is not None:
            data[attr] = value
    return data
```

After that fix, you don't need to report `ATTR_CHANGED_BY` again here.

Can you open a separate PR with that fix? After that is merged you can rebase on latest dev.

Reply:
Here's the PR for that: https://github.com/home-assistant/home-assistant/pull/12049